Feat: Add support of multiple datasets in config

EvilFreelancer opened the pull request:
#### Context

What is the purpose of this PR? Is it to
- [x] add a new feature
- [ ] fix a bug
- [x] update tests and/or documentation
- [ ] other (please add here)

Please link to any issues this PR addresses.

#### Changelog

I've added the ability to use multiple sources for any types of datasets.

After the merger of this PR, users of TorchTune will be able to pass multiple datasets in different formats. For example, it will be possible to mix chat and instruct datasets using different templates, splits, etc.

Example of the new version of the config:

```yml
dataset:
  - _component_: torchtune.datasets.instruct_dataset
    source: tatsu-lab/alpaca
    template: AlpacaInstructTemplate
    split: train
    train_on_input: True
  - _component_: torchtune.datasets.chat_dataset
    source: Open-Orca/SlimOrca-Dedup
    conversation_style: sharegpt
    chat_format: Llama2ChatFormat
    max_seq_len: 1024
    split: train
seed: null
shuffle: True
```

For backward compatibility, users can continue using the previous format of the `dataset` field if they do not wish to use multiple datasets:

```yaml
dataset:
  _component_: torchtune.datasets.instruct_dataset
  source: tatsu-lab/alpaca
  template: AlpacaInstructTemplate
  split: train
  train_on_input: True
```

To run unit tests:

```shell
pytest ./tests/torchtune/datasets/test_multi_dataset.py
```

#### Test plan

Please make sure to do each of the following if applicable to your PR. (If you're not sure about any one of these just ask and we will happily help.)

- [x] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [x] add unit tests for any new functionality
- [x] update docstrings for any new or updated methods or classes
- [x] run unit tests via `pytest tests`
- [x] run recipe tests via `pytest tests -m integration_test`
- [x] manually run any new or modified recipes with sufficient proof of correctness
	- [x] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

EvilFreelancer commented:
Hi! I've signed CLA, how to rerun failed check?

RdoubleA commented:
Thanks for this awesome PR @EvilFreelancer! Handling multiple datasets has been a north star for us, as most mature data pipelines for fine-tuning models typically incorporate multiple data sources, so I appreciate you adding this.

> The only one problem is: dataset must have simmilar formats.

This is my main concern - the concatenated datasets must have the same columns AND the same instruct template / chat format. Also, all the keyword arguments will need to be shared (max_seq_len, train_on_input). This is quite restrictive and will require users to do a lot of offline preprocessing work. Ideally, using multiple datasets should be flexible enough that each dataset can have different columns and different template yet we're able to coalesce these together. In the end, all the data gets tokenized as `Message`s, so I do think this is possible.

What are your thoughts on letting `InstructDataset` and `ChatDataset` handle single data sources only, and creating some container class that can hold multiple `InstructDatasets` or `ChatDatasets`? Since the dataset classes return the token IDs ready to be used by the model, the container class can focus on handling the logic of sampling from the list of datasets. 
```
class MultiDataset(torch.utils.data.Dataset):
    def __init__(self, datasets: List[Dataset]):
        self.datasets = datasets

    def __getitem__(self, index: int) -> Tuple[List[int], List[int]]:
        # Figure out how to sample/interleave multiple datasets here
```
Then, in the config, you could specify multiple datasets like this (or something similar):
```
dataset:
  - _component_: torchtune.datasets.instruct_dataset
    source: tatsu-lab/alpaca
    ...
  - _component_: torchtune.datasets.instruct_dataset
    source: vicgalle/alpaca-gpt4
    ...
```
And the recipe can instantiate a MultiDataset if the dataset param is a list. This way, each dataset can keep their individual parameters.

Ideally, we can make use of `concatenate_datasets` and `interleave_datasets` from HF but since we have preprocessing logic in our dataset classes that are specific to each data source, we might have to create similar logic ourselves.

EvilFreelancer commented:
Hi @RdoubleA, thank you for your response!

Hm, the `MultiDataset` class and the possibility to pass an array of datasets through the `dataset` parameter - it's a great idea, and it's much simpler for end-users to understand how to use this feature than dealing with datasets in similar formats.

I have a couple of spare days, so I may start working on implementing this solution, as it is critical for the project I am working on to have the ability to train on various combinations of datasets.

EvilFreelancer commented:
Hi! I've implemented logic of [MultiDataset](https://github.com/pytorch/torchtune/pull/889/files#diff-404ae02c70619a365d896837164b187a3b5e89ca5b58a5eb73d3de0c1445559b) and enabled it in traning recipes.

I've also removed original multi-source logic and cleaned tests.

How to use:
```yml
dataset:
  - _component_: torchtune.datasets.instruct_dataset
    source: tatsu-lab/alpaca
    template: AlpacaInstructTemplate
    split: train
    train_on_input: True
  - _component_: torchtune.datasets.instruct_dataset
    source: vicgalle/alpaca-gpt4
    template: AlpacaInstructTemplate
    split: train
    train_on_input: True
  - _component_: torchtune.datasets.chat_dataset
    source: Open-Orca/SlimOrca-Dedup
    conversation_style: sharegpt
    chat_format: Llama2ChatFormat
    max_seq_len: 1024
    split: train
  - _component_: torchtune.datasets.chat_dataset
    source: ajibawa-2023/Code-290k-ShareGPT
    conversation_style: sharegpt
    chat_format: Llama2ChatFormat
    max_seq_len: 1024
    split: train
seed: null
shuffle: True
```

By the way, this gave an interesting side effect: using this logic, you can freely mix datasets of any format.

In the example that I gave, dataset formats `instrcut` and `chat` are mixed, it seems to me that this will be a very interesting and convenient feature of TorchTune.

RdoubleA commented on recipes/full_finetune_distributed.py:
```diff
@@ -357,10 +357,12 @@ def _setup_data(
-            tokenizer=self._tokenizer,
-        )
+
+        if isinstance(cfg_dataset.get(0), DictConfig):
```
you can also just directly check if it's a ListConfig. If it's a single dataset then this might fail
```suggestion
        if isinstance(cfg_dataset, ListConfig):
```

I also wonder if there's a better way to handle this so we don't have to repeat this if-else check across all recipes...

EvilFreelancer replied:
Good point, i've replaced it with `ListConfig` check.

EvilFreelancer replied:
> I also wonder if there's a better way to handle this so we don't have to repeat this if-else check across all recipes...

Yeah, this logic can be moved to `config.instantiate` method, but I guess it will break single responsibility principle.

RdoubleA commented on tests/torchtune/datasets/test_alpaca_dataset.py:
Do we still need to make these changes to the test samples to use Dataset across all the individual dataset test files?

EvilFreelancer replied:
I think we need to keep the `datasets.Dataset` objects, since this is a common format and anyone can see that the data should come in this format, plus there is no need to perform any extra transformations, as was the case, for example in `_chat` tests.

RdoubleA commented on torchtune/utils/multi_dataset.py:
```diff
@@ -0,0 +1,46 @@
+
+
+class MultiDataset(Dataset):
+    def __init__(self, datasets: DictConfig[DictConfig], tokenizer):
```
I'm not sure about passing in the datasets as config objects, because then this class is hard to use without having to create a config. And if you ever want to play around with `MultiDataset` in a notebook environment, this might get tedious. Config objects should be separate from the actual code as much as possible, except in the recipe where components need to be instantiated. But I see that it's hard to create a `MultiDataset` without having to instantiate the individual datasets first...

What do you think about an alternate approach where we instantiate the datasets outside of this class and then pass in the dataset objects?
```
# in recipe
datasets = [config.instantiate(single_cfg_dataset, tokenizer) for single_cfg_dataset in cfg_dataset]
ds = MultiDataset(datasets=datasets)

# Modified MultiDataset constructor - won't need tokenizer anymore
def __init__(self, datasets: List[Dataset]):
```

EvilFreelancer replied:
Clever idea, patch added.

RdoubleA commented on torchtune/utils/multi_dataset.py:
```diff
@@ -0,0 +1,46 @@
+
+        # Load all datasets one by one
+        for dataset in datasets:
+            self._data.extend(self._loading_dataset(dataset))
```
This means that all the data will iterate sequentially, right? What if I wanted to interleave the datasets and jump around during training instead of exhausting one dataset at time? Would the `DistributedSampler` handle this?

EvilFreelancer replied:
Yeah, you right, i've changed this logic a new one, please refer MultiDataset class.

RdoubleA commented on torchtune/utils/multi_dataset.py:
```diff
@@ -0,0 +1,46 @@
+
+        output = []
+        for tokens, labels in tqdm(items):
+            output.append((tokens, labels))
```
this is nice because it will do all the templating and tokenization for all your datasets before we even start training so training might be a bit faster, but I imagine users might want to have control over this, in case they want to do the sample preparation / templating on the fly instead (like when you're just debugging things or you have very large datasets, you don't want to process all the datasets in their entirety every time)

EvilFreelancer replied:
Indeed, replaced to kind of `sample preparation`.

RdoubleA reviewed (commented):
I like the simplicity of the `MultiDataset` class you implemented, very easy to understand! There's some considerations around how we instantiate each individual dataset we should figure out.

EvilFreelancer commented:
@RdoubleA hi! Thanks for your review, requested fixes added.

By the way I've noticed one small thing, for example in `vicgalle/alpaca-gpt4` dataset have 52k rows, why TorchTune shows me 26k items on training stage? Maybe it's because train/val splits?

RdoubleA commented on torchtune/utils/multi_dataset.py:
```diff
@@ -0,0 +1,34 @@
+            cumulative_index += len(dataset)
+
+    def __len__(self) -> int:
+        return sum(len(sublist) for sublist in self._datasets)
```
let's just store this on init so we don't have to calculate it every time

EvilFreelancer replied:
Thanks, fixed

RdoubleA commented on torchtune/utils/multi_dataset.py:
```diff
@@ -0,0 +1,34 @@
+
+    def __getitem__(self, index: int) -> Tuple[List[int], List[int]]:
+        if index < 0 or index >= self.__len__():
+            raise IndexError("Index out of range")
```
hm, will this cause negative indexing into MultiDataset to fail? but if you support negative indexing then it might complicate the logic below. Maybe it's fine to restrict it.

EvilFreelancer replied:
I guess you right, it's probably not possible to out of range, i've removed this check.

RdoubleA commented on torchtune/utils/multi_dataset.py:
```diff
@@ -0,0 +1,34 @@
+        self._datasets = datasets
+
+    def __getitem__(self, index: int) -> Tuple[List[int], List[int]]:
+        if index < 0 or index >= self.__len__():
```
nit: reads a bit better
```suggestion
        if index < 0 or index >= len(self):
```

RdoubleA reviewed (commented):
This is turning into something really nice! Thanks for your patience working through this. Could you add a unit test for `MultiDataset` on some toy data?

Also tagging @ebsmothers, @kartikayk to get their thoughts on how we should handle instantiating this in the recipes. I think we could just make every dataset a `MultiDataset`, since you could just pass a list with a single dataset and it should work the same (correct me if I'm wrong here). That should simplify the logic in the recipes.

EvilFreelancer commented:
@RdoubleA hi! I've added fixes you mentioned, and couple simple test on `MultiDataset` class.

What do you think about:

```python
if not isinstance(cfg_dataset, ListConfig):
    cfg_dataset = [cfg_dataset]
datasets = [config.instantiate(cfg_item, tokenizer=self._tokenizer) for cfg_item in cfg_dataset]
ds = utils.MultiDataset(datasets=datasets)
```

instead of:

```
if isinstance(cfg_dataset, ListConfig):
    datasets = [
        config.instantiate(single_cfg_dataset, tokenzier=self._tokenizer)
        for single_cfg_dataset in cfg_dataset
    ]
    ds = utils.MultiDataset(datasets=datasets)
else:
    ds = config.instantiate(cfg_dataset, tokenizer=self._tokenizer)
```

RdoubleA commented:
@EvilFreelancer Thanks for the updates! I'll take another pass soon.

In the meantime, since this is a significant feature we are adding, I'd like to make sure this is rigorously tested. Do you mind confirming the following:
- The unit test for multidataset passes (please add the test command to the PR summary)
- Run one of the distributed recipes and confirm that loss curves and tokens/sec look reasonable (you can use WandBLogger for easy visualization). Main thing I want to confirm here is that we can still sample from multiple datasets correctly in a distributed environment
- Run one of the single device recipes for a similar reason above

Also, I forgot to mention this earlier, but I think a better location for `MultiDataset` would be in `torchtune/datasets` instead of `torchtune/utils`, what do you think?

ebsmothers commented on torchtune/utils/multi_dataset.py:
```diff
@@ -0,0 +1,39 @@
+
+
+class MultiDataset(Dataset):
+    def __init__(self, datasets: List[List[Dataset]]):
```
Why List[List[ here?

EvilFreelancer replied:
Hi, oops, it's a typo from the first iterations of the class, changed to `List[Dataset]` as it should be, thanks!

ebsmothers commented on torchtune/utils/multi_dataset.py:
```diff
@@ -0,0 +1,39 @@
+        # Calculate distribution of indexes in all datasets
+        cumulative_index = 0
+        for idx, dataset in enumerate(datasets):
+            next_cumulative_index = cumulative_index + len(dataset)
```
Not a blocker for this PR, but worth thinking about how we're gonna do this for iterable datasets @RdoubleA @SLR722

ebsmothers commented on torchtune/utils/multi_dataset.py:
```diff
@@ -0,0 +1,39 @@
+        for start, stop, dataset_index in self._indexes:
+            if start <= index < stop:
+                dataset = self._datasets[dataset_index]
+                return dataset[index - start]  # noqa
```
why is noqa needed here?

EvilFreelancer replied:
In first iteration of MultiDataset on this line was a type check error, but now noqa is not required. Thanks for noticing.

EvilFreelancer commented:
@joecummings hi!

> Can you update the README with examples on how this will now look in the YAML config?

I've [added](https://github.com/pytorch/torchtune/pull/889/files#diff-6f46eaee6237147c156aa798061cbdc81ed1f598387e5d059eb9306132e7ba8c) small section to dataset page in tutorial.

EvilFreelancer commented:
The tests for Multi GPU Gemma 2B training are available [here](https://api.wandb.ai/links/miteigi-nemoto/ewedo1mb).

RdoubleA commented:
@EvilFreelancer Thanks for sharing the test runs! Sorry for the delay in following up on this. An update from our side is that we've been having a lot of discussions around iterable datasets, which is something that we want to support more and design around moving forward. This would enable more flexibility around interleaving, weighted sampling, etc. for multiple datasets that won't fit in memory.

That being said, I still think the `MultiDataset` here is valuable for datasets that fit in memory and can still leverage map-style functionality. This will unblock users that want to quickly concatenate multiple data sources until we add iterable datasets and a more powerful `MultiDataset`. So to accurately reflect the scope of this class, my suggestion is to rename this to `ConcatDataset` (you might've started with this name to begin with, I apologize) and make it clear in the docstrings that this is for map datasets that all fit in memory.

Let me know if this makes sense!

RdoubleA commented on torchtune/datasets/_multi.py:
```diff
@@ -0,0 +1,39 @@
+log = utils.get_logger("DEBUG")
+
+
+class MultiDataset(Dataset):
```
Let's add a comprehensive docstring here

EvilFreelancer replied:
A comprehensive docstring added.

ebsmothers commented on docs/source/tutorials/datasets.rst:
```diff
@@ -47,6 +47,30 @@ You could tweak :code:`max_seq_len` to achieve that directly from the config.
     # Original is 512
     max_seq_len: 256
 
+If you wish to train your model on multiple datasets and apply specific settings to each
+one, such as template, maximum sequence length, or chat format, you will need to adopt a
+different configuration format. This approach allows for a fine-grained customization of
+each dataset included in the training. Here is an example that demonstrates this flexibility:
+
+.. code-block:: yaml
+
+  # Dataset
+  dataset:
+    - _component_: torchtune.datasets.instruct_dataset
+      source: vicgalle/alpaca-gpt4
+      template: AlpacaInstructTemplate
+      split: train
+      train_on_input: True
+    - _component_: torchtune.datasets.instruct_dataset
+      source: samsum
+      template: SummarizeTemplate
+      column_map: {"output": "summary"}
+      split: train
+      train_on_input: False
+
+This configuration demonstrates how to seamlessly integrate two datasets from different
+formats for comprehensive training.
```
Can we remove this from the live docs? I feel like there are too many subtleties here that are not immediately clear. E.g. a very natural next question based on this is "how can I sample from different datasets with different probabilities?", which is not supported. So I think it's better to leave it out until we have a full feature set, then we can showcase the functionality properly and in more detail.

EvilFreelancer replied:
Maybe better just to change it to something like this?

```yml
  # Dataset
  dataset:
    - _component_: torchtune.datasets.alpaca_dataset
      max_seq_len: 256
    - _component_: torchtune.datasets.samsum_dataset
```

ebsmothers replied:
So actually my comment is less on the config itself, but more on the way that we frame our multi-dataset support. Basically I want to make sure we are explicit about what's supported and what's not. So while the ability to configure each individual dataset separately does imply some flexibility, I would prefer not to focus on flexibility as much as you have in this particular snippet, as it can easily imply sampling strategies etc. that we do not (yet) support. Instead it'd be good to clearly state that we are concatenating the datasets when built in this way. How about something like this?

```suggestion
It is also possible to train on multiple datasets by combining them into a single :class:`~torchtune.datasets.ConcatDataset`. For example:
.. code-block:: yaml

  dataset:
    - _component_: torchtune.datasets.instruct_dataset
      source: vicgalle/alpaca-gpt4
      template: AlpacaInstructTemplate
      split: train
      train_on_input: True
    - _component_: torchtune.datasets.instruct_dataset
      source: samsum
      template: SummarizeTemplate
      column_map: {"output": "summary"}
      split: train
      train_on_input: False

The preceding snippet demonstrates how you can configure each individual dataset's parameters, then combine them into a single concatenated dataset for training.
```

(Note: for the above suggestion to work you will also need to add `ConcatDataset` to [api_ref_datasets.rst](https://github.com/pytorch/torchtune/blob/9274c8989cfa75feb8804a8e15b5c8ba9d71ccdb/docs/source/api_ref_datasets.rst#L4))

EvilFreelancer replied:
I like it!

Thank you for pointing out `api_ref_datasets.rst`, I will address it promptly.

EvilFreelancer replied:
`ConcatDataset` added to `api_ref_datasets.rst` as you suggested.

ebsmothers commented on recipes/full_finetune_distributed.py:
```diff
@@ -357,10 +357,16 @@ def _setup_data(
-            tokenizer=self._tokenizer,
-        )
+
+        if isinstance(cfg_dataset, ListConfig):
+            datasets = [
+                config.instantiate(single_cfg_dataset, tokenizer=self._tokenizer)
+                for single_cfg_dataset in cfg_dataset
+            ]
```
Not necessary for this PR, but it'd be nice to have a utility like 

```
def instantiate_list_config(cfg: ListConfig, **kwargs) -> List[Any]:
    return [
        instantiate(element, kwargs) for element in cfg
    ]
```

to reduce copypasta across the recipes a bit. cc @RdoubleA to tackle in a follow-up

ebsmothers commented on tests/torchtune/datasets/test_alpaca_dataset.py:
```diff
@@ -28,17 +29,19 @@ def test_label_no_masking(self, load_dataset, tokenizer):
-                ),
-            }
-        ]
+        load_dataset.return_value = Dataset.from_list(
```
What's the story with these `Dataset.from_list` changes? I know it will give us the right return type (`Dataset` instead of raw `List`), anything besides that motivating the change? (I am fine with keeping them in, mainly asking out of curiosity)

RdoubleA replied:
Technically it's a more accurate return type of load dataset, I think it's ok to just leave these out and stick to primitives for simplicity, but I don't have a strong opinion here.

EvilFreelancer replied:
I've invested considerable time into understanding how to test my new dataset class. It was initially unclear regarding the required format and content of its elements. Hence, I suggest providing clear guidelines to save fellow programmers time, elucidating the expected format for dataset elements.

ebsmothers replied:
Thanks, this is very good feedback. I agree that the contracts of various dataset components are not always obvious and take some time to sort through. Aside from improving live docs and better code comments, I'm open to any suggestions you have on how to make this clearer based on your experience.

ebsmothers commented:
Ah crap one more thing: my suggestion screwed up the rendering of the multi-dataset YAML. Sorry about that! I think you just need to add back a newline before the codeblock statement.

<img width="897" alt="Screenshot 2024-05-03 at 2 35 18 PM" src="https://github.com/pytorch/torchtune/assets/24319399/00eb7c23-9fac-4f68-a2ab-e1661f7c1e13">

EvilFreelancer commented:
@ebsmothers typo fixed

ebsmothers reviewed (approved):
Looks great, thank you for adding this!

EvilFreelancer commented:
Thank you for the meticulous code-review and excellent advice, I thoroughly enjoyed working on this project with all of you!
